src/plugins/controls/public/options_list/components/options_list_popover_suggestions.tsx

```diff
@@ -87,53 +102,90 @@ export const OptionsListPopoverSuggestions = ({
           ) : undefined,
       };
     });
-    const suggestionsSelectableOptions = existsSelectableOption
-      ? [existsSelectableOption, ...options]
-      : options;
-    setSelectableOptions(suggestionsSelectableOptions);
+
+    if (canLoadMoreSuggestions) {
+      options.push({
+        key: 'loading-option',
+        className: 'optionslist--loadingMoreGroupLabel',
+        label: OptionsListStrings.popover.getLoadingMoreMessage(),
+        isGroupLabel: true,
+      });
+    } else if (options.length === MAX_OPTIONS_LIST_REQUEST_SIZE) {
+      options.push({
+        key: 'no-more-option',
+        className: 'optionslist--endOfOptionsGroupLabel',
+        label: OptionsListStrings.popover.getAtEndOfOptionsMessage(),
+        isGroupLabel: true,
+      });
+    }
+    setSelectableOptions(existsSelectableOption ? [existsSelectableOption, ...options] : options);
   }, [
     suggestions,
     availableOptions,
     showOnlySelected,
     selectedOptionsSet,
     invalidSelectionsSet,
     existsSelectableOption,
+    canLoadMoreSuggestions,
   ]);
 
+  const loadMoreOptions = useCallback(() => {
+    const listbox = listRef.current?.querySelector('.euiSelectableList__list');
+    if (!listbox) return;
+
+    const { scrollTop, scrollHeight, clientHeight } = listbox;
+    if (scrollTop + clientHeight >= scrollHeight - parseInt(euiThemeVars.euiSizeXXL, 10)) {
+      // reached the "bottom" of the list, where euiSizeXXL acts as a "margin of error" so that the user doesn't
+      // have to scroll **all the way** to the bottom in order to load more options
+      loadMoreSuggestions(totalCardinality ?? MAX_OPTIONS_LIST_REQUEST_SIZE);
+    }
+  }, [loadMoreSuggestions, totalCardinality]);
+
+  useEffect(() => {
+    const container = listRef.current;
+    if (!isLoading && canLoadMoreSuggestions) {
+      container?.addEventListener('scroll', loadMoreOptions, true);
+      return () => {
```

Comment:
Have you tested that this is called every time you expect it to be called? I have no reason to believe this _wouldn't_ work, I just haven't often seen `useEffect` cleanup functions being returned inside an `if` statement.

Reply:
Yup, I tested this pretty thoroughly when I first added it (and just double checked by putting a print statement for both when the event listener is added and when it is destroyed). It seems to work exactly as expected :+1: